Introduce output data format versioning #2653

AyanSinhaMahapatra opened the pull request:
Add output data format version numbers to the headers and version help
text. Introduce new command line option to switch to the new experimental
data format.

Signed-off-by: Ayan Sinha Mahapatra <ayansmahapatra@gmail.com>

AyanSinhaMahapatra committed:
Introduce output data format versioning #2653

Add output data format version numbers to the headers and version help
text. Introduce new command line option to switch to the new experimental
data format.

Signed-off-by: Ayan Sinha Mahapatra <ayansmahapatra@gmail.com>

AyanSinhaMahapatra committed:
Modify tests to ignore headers

Modifies tests to ignore addition of output data format in headers.

Signed-off-by: Ayan Sinha Mahapatra <ayansmahapatra@gmail.com>

AyanSinhaMahapatra committed:
Bump commoncode to have header changes

Signed-off-by: Ayan Sinha Mahapatra <ayansmahapatra@gmail.com>

AyanSinhaMahapatra committed:
Merge branch 'develop' into 2653-output-data-format-version

Signed-off-by: Ayan Sinha Mahapatra <ayansmahapatra@gmail.com>

AyanSinhaMahapatra committed:
Rename flag name to --future-format

Renames flag to --future-format and updates help text test.
Adds future format version.

Signed-off-by: Ayan Sinha Mahapatra <ayansmahapatra@gmail.com>

AyanSinhaMahapatra committed:
Update more tests to ignore headers

Signed-off-by: Ayan Sinha Mahapatra <ayansmahapatra@gmail.com>

AyanSinhaMahapatra commented:
@pombredanne could you look into this why the tests with `latest_from_pip` are failing like [here](https://dev.azure.com/nexB/scancode-toolkit/_build/results?buildId=4557&view=logs&j=9a634e43-2dfe-534b-56d2-0b7ffad252eb&t=28d573b5-74f3-578a-960b-e035b88f9a42). These are failing as `bin/pip install --force-reinstall --upgrade -e .[dev]` replaces commoncode `21.8.31` with `21.8.27`, which doesn't have the header changes, and hence the failures. 

If I change [this](https://github.com/nexB/scancode-toolkit/blob/develop/setup.cfg#L62) to `21.8.31`, this happens in the [test](https://dev.azure.com/nexB/scancode-toolkit/_build/results?buildId=4563&view=logs&j=9a634e43-2dfe-534b-56d2-0b7ffad252eb&t=28d573b5-74f3-578a-960b-e035b88f9a42&l=50):

```
ERROR: Could not find a version that satisfies the requirement commoncode>=21.8.31 (from scancode-toolkit[dev]) (from versions: 20.9, 20.9.30, 20.10, 20.10.8, 20.10.20, 21.1.14, 21.1.21, 21.4.28, 21.5.12, 21.5.25, 21.6.10, 21.6.11, 21.7.23, 21.8.27)
ERROR: No matching distribution found for commoncode>=21.8.31
```
What am I doing wrong here

AyanSinhaMahapatra committed:
Bump commoncode in setup.cfg

Signed-off-by: Ayan Sinha Mahapatra <ayansmahapatra@gmail.com>

pombredanne commented on src/scancode/cli.py:
```diff
@@ -249,6 +251,13 @@ def validate_depth(ctx, param, value):
 
+@click.option('--future-format',
+    is_flag=True,
+    help='Output the next experimental data format, for JSON and YAML output.'
```
Use future and not experimental, here and elsewhere

pombredanne commented on src/scancode_config.py:
```diff
@@ -75,6 +75,13 @@ def _create_dir(location):
+
+# See https://github.com/nexB/scancode-toolkit/issues/2653 for more information
+# on the data format version
+__output_format_version__ = '1.1'
```
let's use 1.0

pombredanne commented:
> 21.8.31

You need to push 21.8.31 to PyPI too IMHO

pombredanne reviewed (changes requested):
IMHO you could have fewer changes in the diffs if `def check_json_scan(..., ignore_headers=True)` was updated to be  `def check_json_scan(..., include_headers=False)` this way few changes are needed in the tests overall... and only a few tests that test headers will need fixing.

AyanSinhaMahapatra commented:
> You need to push 21.8.31 to PyPI too IMHO

You have to add me at https://pypi.org/project/commoncode/ btw

AyanSinhaMahapatra committed:
Update Output Version to proper SemVer

Signed-off-by: Ayan Sinha Mahapatra <ayansmahapatra@gmail.com>

AyanSinhaMahapatra committed:
Replace ignore_header with check_header attribute

Replaces ignore_header with check_header attribute with default False,
to minimise diff and ignore headers by default.

Signed-off-by: Ayan Sinha Mahapatra <ayansmahapatra@gmail.com>

AyanSinhaMahapatra committed:
Remove --future-format flag

See https://github.com/nexB/scancode-toolkit/issues/2653#issuecomment-913708258

Signed-off-by: Ayan Sinha Mahapatra <ayansmahapatra@gmail.com>

AyanSinhaMahapatra committed:
Update release docs with output data versioning

Signed-off-by: Ayan Sinha Mahapatra <ayansmahapatra@gmail.com>

pombredanne reviewed (approved):
LGTM. Merging!
